### PR TITLE
Simplify structured parallelism interface with modules and runtime.

### PR DESCRIPTION
**Pull request by gbtitus**
Simplify the interfaces for managing structured parallelism in the
module code and runtime.  Specifically, remove the "process task list"
and "free task list" steps, leaving us with just "add to task list" and
"execute tasks in list".  In doing so, also change the task list data
object manipulated and used by the latter two from a type known to the
compiler to just an opaque pointer.

The major functional work here is to change fifo tasking to so that it
maintains its needed task list within the task descriptors in the pool
rather than as a separate data structure.  (Side note: it needs this
task list because it cannot multiplex tasks on threads.  None of the
other tasking layer implementations need or use such a list.)

This removes three primitives, one each type and var pre-declared by the
compiler, and two procs from each locale model and runtime tasking layer
implementation.

### CODE REVIEW

**Comment by gbtitus** [comment]
@ronawho , would you please review this one?

File: compiler/passes/parallel.cpp [hunk 1]
```diff
@@ -375,7 +375,7 @@ bundleArgs(CallExpr* fcall, BundleArgsFnData &baData) {
 
     // Now get the taskList field out of the end count.
```
Comment (ronawho) [inline compiler/passes/parallel.cpp]: out of date comment?

Comment (gbtitus) [inline compiler/passes/parallel.cpp]: No the comment is still up-to-date.  There is still a task list, at least fifo maintains one, but it's not in a separate data structure in that tasking layer, and the compiler doesn't need to know anything about its type. Outside of the runtime all that need be known is that it must be unique for each parallel construct and that it must be big enough to hold a pointer.

File: modules/internal/ChapelBase.chpl [hunk 2]
```diff
@@ -1539,7 +1529,7 @@ module ChapelBase {
   pragma "no doc"
   inline proc _defaultOf(type t) where t: _sync_aux_t return _nullSyncVarAuxFields;
   pragma "no doc"
-  inline proc _defaultOf(type t) where t == _task_list return _nullTaskList;
+  inline proc _defaultOf(type t) where t == c_void_ptr return _nullVoidPtr;
```
Comment (ronawho) [inline modules/internal/ChapelBase.chpl]: Isn't the c_void_ptr _defaultOf already defined at https://github.com/chapel-lang/chapel/blob/master/modules/internal/CPtr.chpl#L94-L96

**Comment by ronawho** [comment]
Module, compiler, testing, and qthreads code all looks great to me (nice cleanup!)

I had a hard time trying to carefully review the fifo changes. I think I would need to review the code as a whole instead of just the diff, but I'm not sure if that's worthwhile. If you're happy with that code, then I don't feel the need to review in depth (but of course I'm happy to spend more time reviewing that if you'd like)

File: modules/internal/ChapelBase.chpl (hunk 2 above)
Comment (gbtitus) [inline modules/internal/ChapelBase.chpl]: Ah, didn't spot that. Things seem to work fine with this line removed, so I'll do another commit to take it out.

**Comment by gbtitus** [comment]
I'm happy with it and think it's been tested pretty well. (I'm doing a last bit of testing now, examples with comm=gasnet on chapcs00, with that excess line you pointed out removed.) Unless you really want to slog through it I'm willing to take the heat if anything breaks when it's merged. Oh, wait, I'd have to do that anyway.   :-)

**Comment by ronawho** [comment]
So long as you promise to never leave, I don't need to slog through it  : )

**Comment by gbtitus** [comment]
The final round of testing passed.

**Pull request merged**